pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1053 @@
+			}))
+			.map_err(|e| {
+				// Inner error holds finer granularity but should never occur
+				log::debug!("InvestState transition error: {:?}", e);
+				Error::<T>::from(InvestError::Increase)
+			})?;
+		Pallet::<T>::apply_invest_state_transition(who, investment_id, post_state)?;
+
+		Ok(())
+	}
+
+	#[transactional]
+	fn decrease_foreign_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		amount: T::Balance,
+		foreign_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<(), DispatchError> {
+		let pre_state = InvestmentState::<T>::get(who, investment_id).unwrap_or_default();
+
+		// NOTE: This adds one db read but let's be safe for the MVP, can hopefully be
+		// removed before deploying to production
+		let unprocessed_invest_amount = T::Investment::investment(who, investment_id)?;
+		ensure!(
+			unprocessed_invest_amount == pre_state.get_investing_amount(),
+			DispatchError::Corruption
+		);
+
+		let post_state = pre_state
+			.transition(InvestTransition::DecreaseInvestOrder(Swap {
+				currency_in: pool_currency,
+				currency_out: foreign_currency,
+				amount,
+			}))
+			.map_err(|e| {
+				// Inner error holds finer granularity but should never occur
+				log::debug!("InvestState transition error: {:?}", e);
+				Error::<T>::from(InvestError::Decrease)
+			})?;
+		Pallet::<T>::apply_invest_state_transition(who, investment_id, post_state)?;
+
+		Ok(())
+	}
+
+	#[transactional]
+	fn increase_foreign_redemption(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		amount: T::Balance,
+	) -> Result<(), DispatchError> {
+		let pre_state = RedemptionState::<T>::get(who, investment_id).unwrap_or_default();
+
+		// NOTE: This adds one db read but let's be safe for the MVP, can hopefully be
+		// removed before deploying to production
+		let unprocessed_redeem_amount = T::Investment::redemption(who, investment_id)?;
+		ensure!(
+			unprocessed_redeem_amount == pre_state.get_redeeming_amount(),
+			DispatchError::Corruption
+		);
+
+		let post_state = pre_state
+			.transition(RedeemTransition::IncreaseRedeemOrder(amount))
+			.map_err(|e| {
+				// Inner error holds finer granularity but should never occur
+				log::debug!("RedeemState transition error: {:?}", e);
+				Error::<T>::from(RedeemError::Increase)
+			})?;
+		Pallet::<T>::apply_redeem_state_transition(who, investment_id, post_state)?;
+
+		Ok(())
+	}
+
+	#[transactional]
+	fn decrease_foreign_redemption(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		amount: T::Balance,
+	) -> Result<T::Balance, DispatchError> {
+		let pre_state = RedemptionState::<T>::get(who, investment_id).unwrap_or_default();
+
+		// NOTE: This adds one db read but let's be safe for the MVP, can hopefully be
+		// removed before deploying to production
+		let unprocessed_redeem_amount = T::Investment::redemption(who, investment_id)?;
+		ensure!(
+			unprocessed_redeem_amount == pre_state.get_redeeming_amount(),
+			DispatchError::Corruption
+		);
```

Comment:
Same as above. We should just error out if an collect is required. Need to adapt pallet-investments for that IMO.